### PR TITLE
Check for unique ID columns in origins/destinations

### PR DESCRIPTION
**Pull request by wklumpen**
See #140 

I see also somehow that the hash assertion code has stuck around (trying to keep branches separate) but it looks like it's just the parentheses fix is there.

### CODE REVIEW

File: src/r5py/r5/travel_time_matrix_computer.py [hunk 1]
```diff
@@ -45,6 +45,14 @@
 }
 
 
+class NoIDColumnError(Exception):
```
Comment (HTenkanen) [inline src/r5py/r5/travel_time_matrix_computer.py:48]: Not sure but should we perhaps have a separate file for these errors? That's what I've typically done. Thoughts, @christophfink?

File: src/r5py/r5/travel_time_matrix_computer.py [hunk 2]
```diff
@@ -92,10 +100,24 @@ def __init__(
             transport_network = TransportNetwork(*transport_network)
         self.transport_network = transport_network
 
+        # Quick validation to ensure an ID column exists
+        if "id" not in origins.columns:
```
Comment (HTenkanen) [inline src/r5py/r5/travel_time_matrix_computer.py:104]: As discussed in the issue, I would perhaps also add the possibility to specify a column which is used as unique id. It's a bit silly restriction to always expect 'id' column to be found as it can be called anything.

File: src/r5py/util/classpath.py [hunk 3]
```diff
@@ -36,7 +36,7 @@
     R5_CLASSPATH = str(pathlib.Path(CACHE_DIR) / pathlib.Path(R5_JAR_URL).name)
     try:
         with open(R5_CLASSPATH, "rb") as jar:
-            assert hashlib.sha256(jar.read()).hexdigest == R5_JAR_SHA256
+            assert hashlib.sha256(jar.read()).hexdigest() == R5_JAR_SHA256
```
Comment (HTenkanen) [inline src/r5py/util/classpath.py:39]: Shouldn't this be in a separate PR? Did it slip here?

Comment (HTenkanen) [reply src/r5py/util/classpath.py:39]: Ah, I see you commented about this already

**Comment by HTenkanen** [comment]
@wklumpen Looks good, a few comments. Would you also write tests for this unique id issue? Let's try not to get further behind with the tests because we already now have quite a lot of catching up to do.

**Review by HTenkanen — CHANGES_REQUESTED** [review]

**Comment by wklumpen** [comment]
> @wklumpen Looks good, a few comments. Would you also write tests for this unique id issue? Let's try not to get further behind with the tests because we already now have quite a lot of catching up to do.

I can certainly do a test - the problem is loading (or creating) and invalid dataset with a test. I suppose I could generate a simple invalid GeoJSON on the fly?

Maybe a larger question going forward: Should tests be attached to the PRs or added as a separate task?

File: src/r5py/util/classpath.py (hunk 3 above)
Comment (wklumpen) [reply src/r5py/util/classpath.py:39]: Yep. If I could make Git forget all about it I would....

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 1 above)
Comment (wklumpen) [reply src/r5py/r5/travel_time_matrix_computer.py:48]: Happy to do that - suggestions on naming convention/location?

`util/errors.py` perhaps?

Comment (christophfink) [reply src/r5py/r5/travel_time_matrix_computer.py:48]: sounds great! 

Can we: 
- call the file `util/exceptions.py` (I’d expect more to come), and 
- inherit from 
    - a more specific exception (maybe `ValueError`?), **and** 
    - a generic R5pyException (so we have the possibility to catch all r5py errors with `except R5pyException:` )

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 2 above)
Comment (christophfink) [reply src/r5py/r5/travel_time_matrix_computer.py:104]: I guess a separate issue or PR would be a good place :+1:

File: src/r5py/r5/travel_time_matrix_computer.py [hunk 4]
```diff
@@ -92,10 +100,24 @@ def __init__(
...
+            raise NoIDColumnError("Origin dataset must contain an 'id' column.")
+        # And to make sure the column is indeed unique
+        if len(origins.id.unique()) < origins.shape[0]:
```
Comment (christophfink) [inline src/r5py/r5/travel_time_matrix_computer.py:107]: This would be more readable, and probably also better performing: https://pandas.pydata.org/docs/reference/api/pandas.Series.is_unique.html

File: src/r5py/r5/travel_time_matrix_computer.py [hunk 5]
```diff
@@ -92,10 +100,24 @@ def __init__(
...
+            raise NotUniqueIDError("Origin id values must be unique.")
         self.origins = origins
 
         if destinations is None:
             destinations = origins
+
+        # Quick validation to ensure an ID column exists
```
Comment (christophfink) [inline src/r5py/r5/travel_time_matrix_computer.py:114]: Could you take this entire block into a utility function that can be called from `origins()` and from `destinations()`, and possibly be reused in other classes in the future?

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 2 above)
Comment (HTenkanen) [reply src/r5py/r5/travel_time_matrix_computer.py:104]: Yep, sounds good 👍🏻

**Comment by HTenkanen** [comment]
> I can certainly do a test - the problem is loading (or creating) and invalid dataset with a test. I suppose I could generate a simple invalid GeoJSON on the fly?

Yes, you can create a simple dummy dataset to reproduce the issue (i.e. have non-unique values) and then use that to test that r5py handles it correctly.
 
> Maybe a larger question going forward: Should tests be attached to the PRs or added as a separate task?

Typically tests are attached to the same PR, and that's how we also suggest to do in our contribution guide, so let's follow this convention: https://r5py.readthedocs.io/en/latest/CONTRIBUTING.html#making-changes-and-writing-tests

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 5 above)
Comment (christophfink) [reply src/r5py/r5/travel_time_matrix_computer.py:114]: This utility function could be a private method of whatever should be the parent class of all ‘worker classes’, but for the time being a separate file in `utils/` would probably do the job.

**Comment by christophfink** [comment]
> Maybe a larger question going forward: Should tests be attached to the PRs or added as a separate task?
In my opinion, going forward, we should aim to have a test with every new addition of code. Adding a test to existing functionality, then, should be a separate issue+PR?

**Comment by christophfink** [comment]
> I can certainly do a test - the problem is loading (or creating) and invalid dataset with a test. I suppose I could generate a simple invalid GeoJSON on the fly?

I guess just invalidating it in the test would be good? so dropping the id column, renaming the id column, making it non-unique. If that happens in the test it’s clear how the test file was modified, and does not tempt anybody to try to use the corrupt asset? (just my 2c)

**Comment by wklumpen** [comment]
It wouldn't be difficult to generate a dataset in memory during the test in this case (3 rows of data let's say) - while that's not feasible for every test case, should we try and do that approach where possible instead of relying on specific files?

**Comment by christophfink** [comment]
Well, @HTenkanen definitely has a point, when he says that it would be good to have static test datasets that do not change. After all, the same error might affect test data generation and the code that is tested. 

Should we start a new directory with test data in test/data/ or so?

**Comment by wklumpen** [comment]
Not sure why it's failing the tests - trying to import `geopandas` seems to fail, but we definitely require geopandas?

**Comment by HTenkanen** [comment]
> Not sure why it's failing the tests - trying to import `geopandas` seems to fail, but we definitely require geopandas?

@wklumpen : We are not actually requiring geopandas at the moment. It was part of the dev-environment mainly for the purpose of documentation. There was some discussion about the required packages in #52. 

But for real, I think it is only a matter of time when we will have something developed in the source code which will require geopandas, so I'd say it would be okay to add `geopandas` as a dependency for all the environments in `ci/` directory: https://github.com/r5py/r5py/tree/main/ci

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 2 above)
Comment (christophfink) [reply src/r5py/r5/travel_time_matrix_computer.py:97]: Thinking about it, I would argue against a configurable column name. It’s straight-forward for the user to rename a column, at the same time the code to handle dynamic column names necessarily will be fairly bloated. With good documentation, requiring specific column names is a very pythonic way of handling this.

**Comment by christophfink** [comment]
> > Not sure why it's failing the tests - trying to import `geopandas` seems to fail, but we definitely require geopandas?
> 
> @wklumpen : We are not actually requiring geopandas at the moment. It is part of the dev-environment mainly for the purpose of documentation. There was some discussion about the required packages in #52.
> 
> But for real, I think it is only a matter of time when we will have something developed in the source code which will require geopandas (and now we already do need it for tests), so I'd say it would be okay to add `geopandas` as a dependency for all the environments in `ci/` directory: https://github.com/r5py/r5py/tree/main/ci
> 
> Ping @christophfink

Absolutely, `geopandas` is basically necessary to run anything (e.g., some inputs to the `TravelTimeMatrixComputer` are supposed to be `geopandas.GeoDataFrame`s), let’s just add it to the general dependencies. I’ll open a separate issue+PR for this.

**Comment by wklumpen** [comment]
I may have lost the trail on this one - are we just waiting on the weird Windows issues for this PR?

**Comment by christophfink** [comment]
> I may have lost the trail on this one - are we just waiting on the weird Windows issues for this PR?

weirdly, it seems that we would have to, before merging. Does not really mean you cannot push any other changes in the meantime :)

**Comment by wklumpen** [comment]
"I can't do any work I'm on Windows" is the new "my code's compiling"

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 2 above)
Comment (HTenkanen) [reply src/r5py/r5/travel_time_matrix_computer.py:97]: I guess this comment is resolved from this PR. Whether or not to allow configurable column name can be discussed in a separate issue / PR.

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 5 above)
Comment (HTenkanen) [reply src/r5py/r5/travel_time_matrix_computer.py:107]: @christophfink @wklumpen: Let's turn this into a separate issue, so that we can get this PR rolling.

File: src/r5py/r5/travel_time_matrix_computer.py (hunk 4 above)
Comment (HTenkanen) [reply src/r5py/r5/travel_time_matrix_computer.py:100]: Added a separate issue for this in #162

**Comment by HTenkanen** [comment]
@wklumpen and @christophfink, I went through the comments and created a couple of new issues based on some of them that required action, so we can resolve them later.

**Review by HTenkanen — APPROVED** [review]

**Comment by christophfink** [comment]
@HTenkanen this was really not ready to merge! It’s fine to split it into new issues, but this branch did not work as expected, so should not go into main!

**Comment by HTenkanen** [comment]
@christophfink Hmm, okay, it passed all the tests and didn't spot anything problematic, so what is the issue with the PR?

**Comment by christophfink** [comment]
All the unresolved comments went straight into main!

**Comment by christophfink** [comment]
Removing them from this PR would also mean removing the code that added them.
